src/guiguts/tools/pptxt.py

```diff
@@ -0,0 +1,2591 @@
+        be reversed later so obscure Unicode number characters are used that
+        are very unlikely to appear in book texts."""
+        if match_obj.group(2) == "-":
+            repl = "①"
+        elif match_obj.group(2) == "’":
+            repl = "②"
+        elif match_obj.group(2) == "‘":
+            repl = "③"
+        elif match_obj.group(2) == "'":
+            repl = "④"
+        elif match_obj.group(2) == ".":
+            repl = "⑤"
+        elif match_obj.group(2) == "^":
+            repl = "⑥"
+        elif match_obj.group(2) == "—":
+            # Replace emdash with space in this situation
+            repl = " "
+        else:
+            repl = match_obj.group(2)
+        return match_obj.group(1) + repl + match_obj.group(3)
+
+    def get_words_on_line(line: str) -> list[str]:
+        """Parses the text to find all the 'words' on a line
+        and returns them as a (possibly) empty list."""
+
+        if len(line) == 0:
+            return []
+
+        # Words may be surrounded by abandoned HTML tags, entities,
+        # Unicode numeric character references, etc. Remove these
+        # from the line first of all so genuine words are easier to
+        # identify.
+
+        # A document starting with a "<!DOCTYPE>" declaration is
+        # assumed to be an HTML file. We have to treat this HTML
+        # declaration specially as it may overflow onto a second
+        # line in the case of older documents (HTML 4 or XHTML).
+        # Check for this.
+
+        # Variable 'found_long_doctype_declaration' needs to exist
+        # between function calls.
+        nonlocal found_long_doctype_declaration
+
+        if found_long_doctype_declaration:
+            # Second line of a long <!DOCTYPE> declaration. Toss it.
+            found_long_doctype_declaration = False
+            return []
+
+        if re.search(r"<!DOCTYPE", line, re.IGNORECASE) and ">" not in line:
+            # Looks like a two-line <!DOCTYPE declaration. Toss it
+            # and set flag to toss the second line too.
+            found_long_doctype_declaration = True
+            return []
+
+        # A short HTML 5 <!DOCTYPE> declaration will not be detected
+        # by the above tests. However it will be detected as an HTML
+        # 'tag' next and removed from line.
+
+        new_line = ""
+        while res := re.search(r"<(?=[!\/a-z]).*?(?<=[\"A-Za-z0-9\/]|-)>", line):
+            new_line += " " + line[0 : res.start(0)]
+            line = line[res.start(0) + len(res.group(0)) :]
+        if len(line) != 0:
+            new_line += " " + line
+        line = new_line
+
+        # Replace HTML entities (e.g. &amp;) with a space.
+        new_line = ""
+        while res := re.search(r"(&[a-z0-9]+?;)", line):
+            new_line += " " + line[0 : res.start(0)]
+            line = line[res.start(0) + len(res.group(0)) :]
+        if len(line) != 0:
+            new_line += " " + line
+        line = new_line
```

Comment:
If these sections are replacing occurrences of a regular expression with a space, could you use a call to re.sub instead of each block? Something like the following for the HTML entities block (untested)
```
re.sub(r"(&[a-z0-9]+?;)", " ", line)
```

Reply:
Yes. It now does that.